src/transform/src/reduce_elision.rs

```diff
@@ -69,35 +68,77 @@ impl ReduceElision {
 
                         // SumInt32 takes Int32s as input, but outputs Int64s.
                         AggregateFunc::SumInt32 => {
-                            a.expr.clone().call_unary(UnaryFunc::CastInt32ToInt64)
+                            Ok(a.expr.clone().call_unary(UnaryFunc::CastInt32ToInt64))
                         }
 
                         // SumInt64 takes Int64s as input, but outputs numerics.
-                        AggregateFunc::SumInt64 => a
+                        AggregateFunc::SumInt64 => Ok(a
                             .expr
                             .clone()
-                            .call_unary(UnaryFunc::CastInt64ToNumeric(Some(0))),
+                            .call_unary(UnaryFunc::CastInt64ToNumeric(Some(0)))),
 
                         // JsonbAgg takes _anything_ as input, but must output a Jsonb array.
-                        AggregateFunc::JsonbAgg { .. } => MirScalarExpr::CallVariadic {
+                        AggregateFunc::JsonbAgg { .. } => Ok(MirScalarExpr::CallVariadic {
                             func: VariadicFunc::JsonbBuildArray,
                             exprs: vec![a.expr.clone()],
+                        }),
+
+                        // StringAgg takes nested records of strings and outputs a string
+                        AggregateFunc::StringAgg { .. } => match a.expr {
+                            MirScalarExpr::CallVariadic {
+                                func: VariadicFunc::RecordCreate { .. },
+                                ref exprs,
+                            } => {
+                                if let MirScalarExpr::CallVariadic {
+                                    func: VariadicFunc::RecordCreate { .. },
+                                    ref exprs,
+                                } = exprs[0]
+                                {
+                                    Ok(exprs[0].clone())
+                                } else {
+                                    Err(crate::TransformError::Internal(format!(
+                                        "need nested RecordCreate as expr for StringAgg {:?}",
+                                        a.expr,
+                                    )))
+                                }
+                            }
+                            MirScalarExpr::Literal(Ok(ref row), ColumnType { .. }) => {
+                                let explain_failure_str =
+                                    "Expect LIST[LIST[RECORD[STRING, SEPARATOR]]] literal";
+
+                                let datum = row
+                                    .unpack_first()
+                                    .unwrap_list()
+                                    .iter()
+                                    .next()
+                                    .expect(explain_failure_str)
+                                    .unwrap_list()
+                                    .iter()
+                                    .next()
```

Comment:
Is it possible to change all this stuff into an _expression_ that unpacks the doubly-nested record structure at runtime, rather than trying to do it all statically?

That seems safer and less error-prone, and any cost it introduces can probably be eliminated by some optimization pass.

Reply:
Ah, true. This could be rewritten as `record_get(record_get(literal, 0), 0)`, and then it would get evaluated at the `FoldConstants` pass.

Reply:
does this only apply to the literal? Or can I replace the variadic branch with something similar too?

Reply:
it should apply to both branches.

Reply:
it works as expected. Will update the branch